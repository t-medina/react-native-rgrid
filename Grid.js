import '@expo/match-media'
import matchMedia from 'matchmediaquery';
import {Dimensions, StyleSheet} from "react-native";
import isEqual from "lodash.isequal";
import EventEmitter from 'eventemitter3';


const defaults = {
    columns: 12,
    gutterWidth: 24,
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1400
    },
    containerMaxWidths: {
        xs: "100%",
        sm: 540,
        md: 720,
        lg: 960,
        xl: 1140,
        xxl: 1320
    }
};


const FluidClass = "container-fluid";
const RowClass = "row";
const NoActiveColumnClass = "no-column";
const DimensionsUpdated = "DimensionsUpdated";


const matchBreakpoints = (breakpoints) => ({
    xs: true,
    sm: matchMedia("(min-width: " + breakpoints.sm + "px)").matches,
    md: matchMedia("(min-width: " + breakpoints.md + "px)").matches,
    lg: matchMedia("(min-width: " + breakpoints.lg + "px)").matches,
    xl: matchMedia("(min-width: " + breakpoints.xl + "px)").matches,
    xxl: matchMedia("(min-width: " + breakpoints.xxl + "px)").matches,
})

const getBiggestActiveBreakpoint = (breakpoints) => {
    let biggestBreakpoint;

    for (const [breakpoint, active] of Object.entries(breakpoints)) {
        if (active)
            biggestBreakpoint = breakpoint;
        else
            break;
    }

    return biggestBreakpoint;
}

const buildClass = (name, breakpoint, style) => {
    return { name, breakpoint, style };
}


const buildRowClasses = (breakpoints, columns) => {
    const classes = {};
    classes[RowClass] = buildClass(RowClass, "xs", { display: "flex", flexDirection: "row", flexWrap: "wrap" })

    return classes;
}

const buildColumnsClasses = (breakpoints, columns, gutter) => {
    const classes = {};
    const commonStyles  = { paddingLeft: gutter / 2, paddingRight: gutter / 2 };

    Object.keys(breakpoints).forEach(breakpoint => {

        buildSizeColumnClasses(breakpoint);
        buildNoSizeColumnClasses(breakpoint);
        buildAutoWidthClasses(breakpoint);
        buildNoActiveColumnClasses();
    });


    function buildSizeColumnClasses(breakpoint) {
        for (let size = 1; size <= columns; size++)
        {
            const width = Math.round((size / columns * 100) * 100) / 100;
            const style = { width: width + "%", flex: 0, flexBasis: "auto", ...commonStyles };

            const name  = "col-" + (breakpoint === "xs" ? "" : breakpoint + "-") + size;
            classes[name] = buildClass(name, breakpoint, style);
        }
    }

    function buildNoSizeColumnClasses(breakpoint) {
        const name  = "col" + (breakpoint === "xs" ? "" : "-" + breakpoint);
        const style = { flex: 1, flexShrink: 0, ...commonStyles };

        classes[name] = buildClass(name, breakpoint, style);
    }

    function buildAutoWidthClasses(breakpoint) {
        const name  = "col" + (breakpoint === "xs" ? "" : "-" + breakpoint) + "-auto";
        const style = { width: "auto", flex: 0, flexBasis: "auto", ...commonStyles }

        classes[name] = buildClass(name, breakpoint, style);
    }

    function buildNoActiveColumnClasses() {
        const style = { width: "100%", flex: 0, flexBasis: "100%", ...commonStyles }

        classes[NoActiveColumnClass] = buildClass(NoActiveColumnClass, "xs", style);
    }

    return classes;
}

const buildContainersClasses = (breakpoints, containerMaxWidths) => {
    const classes = {};
    const commonStyles  = { marginLeft: "auto", marginRight: "auto" };

    classes[FluidClass] = buildClass(FluidClass, "xs", { width: "100%", ...commonStyles });

    Object.keys(breakpoints).forEach(breakpoint => {
        const name  = "container" + (breakpoint === "xs" ? "" : "-" + breakpoint);
        const style = { width: containerMaxWidths[breakpoint], ...commonStyles };

        classes[name] = buildClass(name, breakpoint, style);
    });

    return classes;
}


class GridDefinition {

    _isReady = false;
    _properties = {};
    _eventEmitter;
    _classes;
    _classesDefinition;
    _activeStyles = {};
    _activeBreakpoints = {};
    _biggestActiveBreakpoint;


    constructor() {
        this._eventEmitter = new EventEmitter();
    }

    build = (config = {}, callback = () => null) => {
        Object.assign(this._properties, defaults, config);

        Dimensions.addEventListener('change', (allDimensions) => {
            if (this._isReady)
                this._updateGrid();
        });

        const rowClasses        = buildRowClasses(this._properties.breakpoints, this._properties.columns);
        const columnsClasses    = buildColumnsClasses(this._properties.breakpoints, this._properties.columns, this._properties.gutterWidth);
        const containersClasses = buildContainersClasses(this._properties.breakpoints, this._properties.containerMaxWidths);

        this._classesDefinition = { ...rowClasses, ...columnsClasses, ...containersClasses };
        this._classes = StyleSheet.create(
            Object.keys(this._classesDefinition)
                  .reduce((ac, current) => ({...ac, [current]: this._classesDefinition[current].style}), {}));

        this._updateGrid();
        this._isReady = true;

        callback();
    }

    getActiveStyles = (classes) => {
        const validClasses = classes.filter(_class => this._classesDefinition[_class] !== undefined);

        const containerClass = validClasses.find(_class => _class.includes("container"));
        if (containerClass !== undefined)
            return this._getActiveContainerStyles(containerClass);

        const rowClass = validClasses.find(_class => _class.includes("row"));
        if (rowClass !== undefined)
            return this._getActiveRowStyles(rowClass);

        const columnClasses = validClasses.filter(_class => _class.includes("col"));
        if (columnClasses.length > 0)
            return this._getActiveColumnStyles(columnClasses);

        return [];
    }

    onDimensionsUpdated(listener) {
        this._eventEmitter.on(DimensionsUpdated, listener);
        return () => this._eventEmitter.off(DimensionsUpdated, listener);
    }


    _updateGrid = () => {
        const matchedBreakpoints = matchBreakpoints(this._properties.breakpoints);

        if (!isEqual(this._activeBreakpoints, matchedBreakpoints)) {
            this._activeBreakpoints = matchedBreakpoints;
            this._biggestActiveBreakpoint = getBiggestActiveBreakpoint(this._activeBreakpoints);
            this._activeStyles = this._getActiveStyles();

            if (this._isReady) {
                this._eventEmitter.emit(DimensionsUpdated);
            }
        }
    }

    _getActiveStyles = () => {
        return Object.values(this._classesDefinition)
                     .filter(({ breakpoint }) => this._activeBreakpoints[breakpoint] === true)
                     .reduce((ac, current) => ({...ac, [current.name]: this._classes[current.name]}), {});
    }

    _getActiveContainerStyles = (containerClass) => {
        if (containerClass === FluidClass || this._biggestActiveBreakpoint === "xs")
            return [ this._classes[FluidClass]];

        const containerBreakpoint = this._classesDefinition[containerClass].breakpoint;
        if (this._properties.breakpoints[containerBreakpoint] > this._properties.breakpoints[this._biggestActiveBreakpoint])
            return [ this._classes[FluidClass] ];

        return [ this._classes["container-" + this._biggestActiveBreakpoint]];
    }

    _getActiveRowStyles = () => {
        return [ this._classes[RowClass] ];
    }

    _getActiveColumnStyles = (columnClasses) => {
        columnClasses.sort((a, b) => {
            const aBreakpoint = this._classesDefinition[a].breakpoint;
            const bBreakpoint = this._classesDefinition[b].breakpoint;

            return this._properties.breakpoints[aBreakpoint] - this._properties.breakpoints[bBreakpoint];
        });

        const activeStyles = columnClasses.filter(responsiveClass => this._activeStyles[responsiveClass] !== undefined)
                                          .map(responsiveClass => this._activeStyles[responsiveClass]);

        return activeStyles.length > 0 ? activeStyles : [ this._classes[NoActiveColumnClass] ];
    }
}


const Grid = new GridDefinition();
export default Grid;
