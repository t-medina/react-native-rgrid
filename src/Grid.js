import '@expo/match-media'
import matchMedia from 'matchmediaquery';
import {Dimensions, StyleSheet} from "react-native";
import EventEmitter from 'eventemitter3';


const defaults = {
    columns: 12,
    gutterWidth: 24,
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1400
    },
    containerMaxWidths: {
        xs: "100%",
        sm: 540,
        md: 720,
        lg: 960,
        xl: 1140,
        xxl: 1320
    }
};

const FluidClass = "container-fluid";
const RowClass = "row";
const NoActiveColumnClass = "no-column";
const DimensionsUpdated = "DimensionsUpdated";


let isReady = false;
let properties = {};
let eventEmitter;
let classes;
let classesDefinition;
let activeStyles = {};
let activeBreakpoints = {};
let biggestActiveBreakpoint;


const matchBreakpoints = (breakpoints) => ({
    xs: true,
    sm: matchMedia("(min-width: " + breakpoints.sm + "px)").matches,
    md: matchMedia("(min-width: " + breakpoints.md + "px)").matches,
    lg: matchMedia("(min-width: " + breakpoints.lg + "px)").matches,
    xl: matchMedia("(min-width: " + breakpoints.xl + "px)").matches,
    xxl: matchMedia("(min-width: " + breakpoints.xxl + "px)").matches,
})

const getBiggestActiveBreakpoint = (breakpoints) => {
    let biggestBreakpoint;

    for (const [breakpoint, active] of Object.entries(breakpoints)) {
        if (active)
            biggestBreakpoint = breakpoint;
        else
            break;
    }

    return biggestBreakpoint;
}


const buildClass = (name, breakpoint, style) => {
    return { name, breakpoint, style };
}

const buildRowClasses = (breakpoints, columns) => {
    const classes = {};
    classes[RowClass] = buildClass(RowClass, "xs", { display: "flex", flexDirection: "row", flexWrap: "wrap" })

    return classes;
}

const buildColumnsClasses = (breakpoints, columns, gutter) => {
    const classes = {};
    const commonStyles  = { paddingLeft: gutter / 2, paddingRight: gutter / 2 };

    Object.keys(breakpoints).forEach(breakpoint => {

        buildSizeColumnClasses(breakpoint);
        buildNoSizeColumnClasses(breakpoint);
        buildAutoWidthClasses(breakpoint);
        buildNoActiveColumnClasses();
    });


    function buildSizeColumnClasses(breakpoint) {
        for (let size = 1; size <= columns; size++)
        {
            const width = Math.round((size / columns * 100) * 100) / 100;
            const style = { width: width + "%", flex: 0, flexBasis: "auto", ...commonStyles };

            const name  = "col-" + (breakpoint === "xs" ? "" : breakpoint + "-") + size;
            classes[name] = buildClass(name, breakpoint, style);
        }
    }

    function buildNoSizeColumnClasses(breakpoint) {
        const name  = "col" + (breakpoint === "xs" ? "" : "-" + breakpoint);
        const style = { flex: 1, flexShrink: 0, ...commonStyles };

        classes[name] = buildClass(name, breakpoint, style);
    }

    function buildAutoWidthClasses(breakpoint) {
        const name  = "col" + (breakpoint === "xs" ? "" : "-" + breakpoint) + "-auto";
        const style = { width: "auto", flex: 0, flexBasis: "auto", ...commonStyles }

        classes[name] = buildClass(name, breakpoint, style);
    }

    function buildNoActiveColumnClasses() {
        const style = { width: "100%", flex: 0, flexBasis: "100%", ...commonStyles }

        classes[NoActiveColumnClass] = buildClass(NoActiveColumnClass, "xs", style);
    }

    return classes;
}

const buildContainersClasses = (breakpoints, containerMaxWidths) => {
    const classes = {};
    const commonStyles  = { marginLeft: "auto", marginRight: "auto" };

    classes[FluidClass] = buildClass(FluidClass, "xs", { width: "100%", ...commonStyles });

    Object.keys(breakpoints).forEach(breakpoint => {
        const name  = "container" + (breakpoint === "xs" ? "" : "-" + breakpoint);
        const style = { width: containerMaxWidths[breakpoint], ...commonStyles };

        classes[name] = buildClass(name, breakpoint, style);
    });

    return classes;
}


const updateGrid = () => {
    const matchedBreakpoints = matchBreakpoints(properties.breakpoints);
    const breakpointsHaveChanged = JSON.stringify(activeBreakpoints) !== JSON.stringify(matchedBreakpoints)

    if (breakpointsHaveChanged) {
        activeBreakpoints = matchedBreakpoints;
        biggestActiveBreakpoint = getBiggestActiveBreakpoint(activeBreakpoints);
        activeStyles = updateActiveStyles();

        if (isReady) {
            eventEmitter.emit(DimensionsUpdated);
        }
    }
}

const updateActiveStyles = () => {
    return Object.values(classesDefinition)
        .filter(({ breakpoint }) => activeBreakpoints[breakpoint] === true)
        .reduce((ac, current) => ({...ac, [current.name]: classes[current.name]}), {});
}


const getActiveContainerStyles = (containerClass) => {
    if (containerClass === FluidClass || biggestActiveBreakpoint === "xs")
        return [ classes[FluidClass]];

    const containerBreakpoint = classesDefinition[containerClass].breakpoint;
    if (properties.breakpoints[containerBreakpoint] > properties.breakpoints[biggestActiveBreakpoint])
        return [ classes[FluidClass] ];

    return [ classes["container-" + biggestActiveBreakpoint]];
}

const getActiveRowStyles = () => {
    return [ classes[RowClass] ];
}

const getActiveColumnStyles = (columnClasses) => {
    columnClasses.sort((a, b) => {
        const aBreakpoint = classesDefinition[a].breakpoint;
        const bBreakpoint = classesDefinition[b].breakpoint;

        return properties.breakpoints[aBreakpoint] - properties.breakpoints[bBreakpoint];
    });

    const activeColumnStyles = columnClasses.filter(responsiveClass => activeStyles[responsiveClass] !== undefined)
        .map(responsiveClass => activeStyles[responsiveClass]);

    return activeColumnStyles.length > 0 ? activeColumnStyles : [ classes[NoActiveColumnClass] ];
}



const build = (config = {}) => {
    eventEmitter = new EventEmitter();

    Object.assign(properties, defaults, config);
    Dimensions.addEventListener('change', (allDimensions) => {
        if (isReady)
            updateGrid();
    });

    const rowClasses        = buildRowClasses(properties.breakpoints, properties.columns);
    const columnsClasses    = buildColumnsClasses(properties.breakpoints, properties.columns, properties.gutterWidth);
    const containersClasses = buildContainersClasses(properties.breakpoints, properties.containerMaxWidths);

    classesDefinition = { ...rowClasses, ...columnsClasses, ...containersClasses };
    classes = StyleSheet.create(
        Object.keys(classesDefinition)
            .reduce((ac, current) => ({...ac, [current]: classesDefinition[current].style}), {}));

    updateGrid();
    isReady = true;
}

const isBuilt = () => {
    return isReady;
}

const getActiveStyles = (classes) => {
    const validClasses = classes.filter(_class => classesDefinition[_class] !== undefined);

    const containerClass = validClasses.find(_class => _class.includes("container"));
    if (containerClass !== undefined)
        return getActiveContainerStyles(containerClass);

    const rowClass = validClasses.find(_class => _class.includes("row"));
    if (rowClass !== undefined)
        return getActiveRowStyles(rowClass);

    const columnClasses = validClasses.filter(_class => _class.includes("col"));
    if (columnClasses.length > 0)
        return getActiveColumnStyles(columnClasses);

    return [];
}

const onDimensionsUpdated = (listener) => {
    eventEmitter.on(DimensionsUpdated, listener);
    return () => eventEmitter.off(DimensionsUpdated, listener);
}

export default {
    build: build,
    isBuilt: isBuilt,
    getActiveStyles: getActiveStyles,
    onDimensionsUpdated: onDimensionsUpdated
};
