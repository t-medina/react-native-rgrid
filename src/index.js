import RGrid from "./components/RGrid";
import RView from "./components/RView";

export { RGrid, RView };
