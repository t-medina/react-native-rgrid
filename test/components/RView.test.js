import React from 'react';
import {Text} from "react-native";
import renderer from 'react-test-renderer';
import Grid from "../../src/Grid";
import {RView} from "../../src";

const mockUnsubscribe = jest.fn();
const mockStyles = {
    "col-12": "style-col-12",
    "col-md-4": "style-col-md-4",
}
jest.mock('../../src/Grid', () => ({
    isBuilt: jest.fn(() => true),
    onDimensionsUpdated: jest.fn(() => mockUnsubscribe),
    getActiveStyles: jest.fn((classes) => {
        return classes.filter(_class => mockStyles[_class]).map(_class => mockStyles[_class]);
    })
}));


describe("RView component render:", () => {
    it("given an empty RView, the component renders successfully", () => {
        const rview = renderer.create(<RView />).toJSON();
        expect(rview).toMatchSnapshot();
    });

    it("given a RView with one child, the component renders successfully", () => {
        const rview = renderer.create(<RView><Text>Sample content</Text></RView>).toJSON();
        expect(rview).toMatchSnapshot();
    });
});


describe("RView Styles:", () => {
    it("given an empty style prop is passed, the component renders without styles", () => {
        const rview = renderer.create(<RView style={{}} />).toJSON();
        expect(rview).toMatchSnapshot();
    });

    it("given an object style prop is passed, the styles are propagated to the generated View", () => {
        const style = { backgroundColor: "#fff" };
        const rview = renderer.create(<RView style={style} />).toJSON();

        expect(rview).toMatchSnapshot();
    });

    it("given an array style prop is passed, the styles are propagated to the generated View", () => {
        const style = [{ backgroundColor: "#fff" }, { fontColor: "#000" }];
        const rview = renderer.create(<RView style={style} />).toJSON();

        expect(rview).toMatchSnapshot();
    });
});


describe("RView and Grid classes:", () => {
    beforeEach(() => {
        Grid.isBuilt = jest.fn(() => true);
        Grid.onDimensionsUpdated.mockClear();
    });

    it("given a RView without RGrid, the renders fails due to the Grid not being built", () => {
        Grid.isBuilt = jest.fn(() => false);
        expect(() => renderer.create(<RView />)).toThrow();
    });

    it("given a RView, the Grid onDimensionsUpdated function is called once", () => {
        renderer.create(<RView />);
        expect(Grid.onDimensionsUpdated).toHaveBeenCalledTimes(1);
    });

    it("given a mounted RView, the unsubscribe function is called once on unmount", () => {
        const rview = renderer.create(<RView />);
        rview.unmount();

        expect(mockUnsubscribe).toHaveBeenCalledTimes(1);
    });

    it("given an empty classes prop is passed, the component renders without styles", () => {
        const rview = renderer.create(<RView classes="" />).toJSON();
        expect(rview).toMatchSnapshot();
    });

    it("given one valid String class, the component retrieves the responsive styles", () => {
        const rview = renderer.create(<RView classes="col-12" />).toJSON();
        expect(rview).toMatchSnapshot();
    });

    it("given a non valid String class, the component renders successfully", () => {
        const rview = renderer.create(<RView classes="invalid-class" />).toJSON();
        expect(rview).toMatchSnapshot();
    });

    it("given a combination of valid and invalid String classes, the component retrieves the responsive styles", () => {
        const rview = renderer.create(<RView classes="col-12 invalid-class col-md-4" />).toJSON();
        expect(rview).toMatchSnapshot();
    });

    it("given one valid String array class, the component retrieves the responsive styles", () => {
        const rview = renderer.create(<RView classes={["col-12"]} />).toJSON();
        expect(rview).toMatchSnapshot();
    });

    it("given a non valid String array class, the component renders successfully", () => {
        const rview = renderer.create(<RView classes={["invalid-class"]} />).toJSON();
        expect(rview).toMatchSnapshot();
    });

    it("given a combination of valid and invalid String array classes, the component retrieves the responsive styles", () => {
        const rview = renderer.create(<RView classes={["col-12", "invalid-class", "col-md-4"]} />).toJSON();
        expect(rview).toMatchSnapshot();
    });
});
