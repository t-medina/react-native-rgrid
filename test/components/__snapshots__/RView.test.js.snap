// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RView Styles: given an array style prop is passed, the styles are propagated to the generated View 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "fontColor": "#000",
      },
    ]
  }
/>
`;

exports[`RView Styles: given an empty style prop is passed, the component renders without styles 1`] = `
<View
  style={
    Array [
      Object {},
    ]
  }
/>
`;

exports[`RView Styles: given an object style prop is passed, the styles are propagated to the generated View 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
    ]
  }
/>
`;

exports[`RView and Grid classes: given a combination of valid and invalid String array classes, the component retrieves the responsive styles 1`] = `
<View
  style={
    Array [
      "style-col-12",
      "style-col-md-4",
    ]
  }
/>
`;

exports[`RView and Grid classes: given a combination of valid and invalid String classes, the component retrieves the responsive styles 1`] = `
<View
  style={
    Array [
      "style-col-12",
      "style-col-md-4",
    ]
  }
/>
`;

exports[`RView and Grid classes: given a non valid String array class, the component renders successfully 1`] = `
<View
  style={Array []}
/>
`;

exports[`RView and Grid classes: given a non valid String class, the component renders successfully 1`] = `
<View
  style={Array []}
/>
`;

exports[`RView and Grid classes: given an empty classes prop is passed, the component renders without styles 1`] = `
<View
  style={Array []}
/>
`;

exports[`RView and Grid classes: given one valid String array class, the component retrieves the responsive styles 1`] = `
<View
  style={
    Array [
      "style-col-12",
    ]
  }
/>
`;

exports[`RView and Grid classes: given one valid String class, the component retrieves the responsive styles 1`] = `
<View
  style={
    Array [
      "style-col-12",
    ]
  }
/>
`;

exports[`RView component render: given a RView with one child, the component renders successfully 1`] = `
<View
  style={Array []}
>
  <Text>
    Sample content
  </Text>
</View>
`;

exports[`RView component render: given an empty RView, the component renders successfully 1`] = `
<View
  style={Array []}
/>
`;
